[CIP 0017] Integrate plumo proofs into the lightest sync client (#41)
celo-org/celo-proposals: CIPs/0017.md

# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol

- Date: 2020-07-27
- Author: @lucasege
- Status: WIP

## Overview

In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.

> The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)

Author: ```suggestion
The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)

More information on Plumo can be found in the [reference paper](https://docs.zkproof.org/pages/standards/accepted-workshop3/proposal-plumo_celolightclient.pdf)
```

## Goals

- Support adding/storing proofs on full nodes

> - Gossip these proofs to propogate them through the network

Author: ```suggestion
- Gossip these proofs to propagate them through the network
```

- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.

## Proposed Solution

We propose to add two new RPCs: 
* `AddProof(proof []byte, firstEpoch uint, lastEpoch uint, versionNumber uint)`
* `Proofs() ([][]byte)`.

`versionNumber` here refers to the snark circuit version, for more info refer to #upgradeability below.

When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propagate the proof throughout the network.

> This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by it's combination of first and last epoch. 

Reviewer: curious why we need to associate the validator set with the verification key -- would the identifier / version number related to the circuit not be sufficient?

Author: This may have been unclear in the wording, trying to make it more clear now.

The version is not related to the validator set, only the VK.

that node A currently *has* and requests those it is missing.
Node B's reply must contain any proofs that node B has that were not included in the message.
If `complement` is false, then this message is sent with the set of proofs that node A currently
*does not have* and requests these from this peer.
Node B's reply must contain the proofs requested, if node B has them. Node A should only request proofs
that node B has indicated via `NewPlumoProof` that it has.

PlumoProofMsg (0x1b)
[plumoProof_0, plumoProof_1, ...]
Reply to GetPlumoProofMsg. The items in the list are plumo proof structs, {proof: []byte, firstEpoch: uint, lastEpoch: uint, versionNumber: uint}.
This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
```

This will likely be accompanied by a new `proofFetcher` object to be in charge of fetching these proofs from peers while providing protections against DoS attacks. This feels like overkill, but seems necessary to prevent attacks and maintain state.

The `les` protocol must then also be extended via at least two messages:
```

> GetCertifiedEpochsValidatorSet - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.

Author: ```suggestion
GetCertifiedEpochsValidatorSet - Sends the current highest epoch known to the server. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
```

## Alternative Solutions

Alternatively, `GetPlumoProofMsg` could utilize a bloom filter instead, allowing for a constant sized message. However, bloom filters allow for false positives: a false positive here would indicate to a peer that the sender has a proof when the sender does not. This could prevent certain proofs from ever getting to some nodes. For example, say node A is on the network and has received a certain amount of proofs via traditional propagation. Node A then goes down for a short period, and within that period proof P is added and propagated. Node A then comes back online, and sends a `GetPlumoProofsMsg` to request any proofs it may have missed, and sends along the corresponding bloom filter for the proofs it holds. If proof P happens to be a false positive for the set of proofs node A already has, then all of node A's peers will simply return nothing. Thus node A will never receive proof P. This can then propagate similarly through the network either through a similar path as node A, or for any nodes where all of their peers have already "lost" proof P.

> `GetPlumoProofMsg` could also request *all* plumo proofs from the receiver, leaving the responsibility of filtering seen proofs with the caller. This is a simpler implementation, but is less scalable as we increase the amount of stored proofs. 

Reviewer: qq -- as a naive reader, I'd be curious, do all proofs have to be stored in perpetuity? Eg. Proof A maps from epoch 0 to epoch 10, proof B from epoch 11-20, incrementing up to epoch 1000. Do we store all 100 proofs?

Wondering why a prover wouldn't aggregate this into one proof from epoch 0-1000?

Author: Not naive at all - this is something we've been mulling over as well.
The circuit that generates these proofs has a "hard-coded" limit of how many epochs a proof can span (I believe right now we're targeting 120*), so while we can't roll up to 1000 like in your example, we could roll up to this limit. 

That said, while we can reduce redundant proofs like you say, currently I'm thinking we shouldn't. The reasons being that `lightest` clients may still ask about blocks within epochs between boundaries, so removing intermediary proofs could hurt the performance there. The other reason is that proofs aren't very large and there likely won't be too many, so the tradeoffs seem worth it here.

* we may be able to change this limit dynamically, but that's not super relevant here

Specific aspects of the `les` changes may have tradeoffs that I've yet to encounter, so any input and feedback on things to consider there would be helpful.

## Risks

Altering the `eth` protocol introduces some risk of introducing bugs, which could inhibit a nodes ability to connect to the network.

> Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain.

Reviewer: Please expand on this and include how it might be prevented?

Also, is there any risk related to producing snarks? Eg with the mpc setup, or if all provers are malicious

Author: There are potential risks inherent to the setup, but these are mitigated via crypto & opsec techniques - will need to ask the folks working on the setup for more details.

Assuming a secure setup and snark construction malicious provers' impact should be mitigated via the same tactics as malicious les servers already are. A malicious prover's proofs should not propagate throughout the network because of the verification step, and their impact on their `les` peers should be eliminated by verification checks once the light client has synced and is receiving blocks from multiple peers. This may also be mitigated earlier via multiple proofs from multiple peers, but I'm not sure on the details yet.

> Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain.

Author: ```suggestion
Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain. Bugs that may be introduced via faulty verification mechanisms will be mitigated via chaos-testing (malicious testers), fuzz testing, e2e tests and strong unit tests.
```

The prover/verifier for Plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)

More information on Plumo can be found in the [reference paper](https://docs.zkproof.org/pages/standards/accepted-workshop3/proposal-plumo_celolightclient.pdf)

## Goals

- Support adding/storing proofs on full nodes
- Gossip these proofs to propagate them through the network
- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly

## Proposed Solution

### Full node changes (RPC and `eth` protocol)
We propose to add two new RPCs: 
* `AddProof(proof []byte, firstEpoch uint, lastEpoch uint, versionNumber uint)`
* `Proofs() ([][]byte)`.

`versionNumber` here refers to the snark circuit version, for more info refer to #upgradeability below.

When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propagate the proof throughout the network.

> This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by the combination of its first and last epoch. 

Author: Yes, I've built it that way just missed updating it here

> This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by the combination of its first and last epoch. 

Author: ```suggestion
This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by the combination of its first and last epoch and the proof's version number. 
```

To propagate these proofs throughout the network, we also propose to extend the `eth` protocol with three new message types: 
```
NewPlumoProof (0x019)
[[firstEpoch, lastEpoch, versionNumber], ...]
Specify one or more new plumo proofs which have appeared on the network.

> This is intended to be honored with a proceeding GetPlumoProofsMsg to request the proof.

Author: ```suggestion
This indicates to other peers that a new proof is available by sending out its metadata. Peers are then expected to send a `GetPlumoProofsMsg` to request this proof by its metadata.
```

If `complement` is false, then this message is sent with the set of proofs that node A currently
*does not have* and requests these from this peer.
Node B's reply must contain the proofs requested, if node B has them. Node A should only request proofs
that node B has indicated via `NewPlumoProof` that it has.

PlumoProofMsg (0x1b)
[plumoProof_0, plumoProof_1, ...]
Reply to GetPlumoProofMsg. The items in the list are plumo proof structs, {proof: []byte, firstEpoch: uint, lastEpoch: uint, versionNumber: uint}.
This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
```

This will be accompanied by a new `proofFetcher` object to be in charge of fetching these proofs from peers while providing protections against DoS attacks. This feels like overkill, but seems necessary to prevent attacks and maintain state.

### `les` protocol changes

Relevant proposed structs:
```
// Same as in the `eth` protocol, minimal proof metadata
ProofMetadata struct {
  FirstEpoch uint
  LastEpoch uint
  VersionNumber uint
}

// Metadata for each epoch to construct `EpochBlock` for the snark's verify fn

> // May be unnecessary if we can make assumptions on `MaxNonSigners`

Author: Yeah good point, if this is easy for clients then agreed it should be done that way.

Still seems we need the `LastEpoch` struct, though

> // May be unnecessary if we can make assumptions on `MaxNonSigners`

Author: Yeah sorry should have removed this.
I was originally wondering if the client could potentially generate this themselves, making this unnecessary, but I believe we need it for the `LastEPoch` still.

> // May be unnecessary if we can make assumptions on `MaxNonSigners`

Author: On the "breaking change" point, this wouldn't introduce any potential for breaking changes, the worst that could happen is some wasted bandwidth (sending something the client doesn't need), but I believe that's already been fixed in the design, I'll update this to reflect that.

> LightEpochBlock struct {

Author: ```suggestion
ProofInputs struct {
```

  Index         uint // Epoch Index 

>   MaxNonSigners uint // MinQuorumSize()

Reviewer 2: We have to set a limit for validator set size in the circuit before doing the trusted setup. It's okay to be under that limit, we pad with dummy public keys. So yes this is fine as long as B is less than this limit.

}

// All of the data needed by the light client to verify and utilize a proof.
// This is also crafted to allow for batch verifications using `FirstHashToField`

> // My rough calculations show this struct is ~439 bytes for a non-batched proof with no val set differences.

Author: Yup it does. Each public key is 96 bytes, so yes the size will be dominated around >4 new validators, the size is roughly `431 + (96 * n)`.

I'm not sure the exact size of the deleted validators bitmap, I'm using 16 bytes as an estimate here which would be 128 bits, but I'm not sure if there is additional necessary overhead or if we need to increase the bits' size.

431 comes from: 383 (proof) + 8 (firstEpochIndex) + 16 (lastEpochBlock) + 8 (versionNumber) + 16 (deletedValBitmap)

LightPlumoProof struct {
  Proof             []byte // Serialized proof. Latest estimate from @psi = 383 bytes
  FirstEpoch        LightEpochBlock // First epoch this proof is over

>   LastEpoch         LightEpochBlock // Last epoch this proof is over

Author: Agreed that this seems to imply that capability, but I can't confirm the prover's capabilities here:
@psivesely do you know?

>   LastEpoch         LightEpochBlock // Last epoch this proof is over

Author: Looks like this is addressed by the comment above, agreed this seems possible

  VersionNumber     uint // Proof version number, indicates the corresponding verification key to use
  LastHashToField  []byte  // Hash to field of Last epoch's `snark.EpochBlock` struct. To be used for batch verification.
                           // If there is no batch verification, this will be empty. May need an additional boolean to signify this.
                           // If batch verification should be used, newValidators and deletedValidators will be empty.

>   NewValidators     []blsCrypto.SerializedPublicKey // ValSetDiff(FirstEpoch, LastEpoch), added validators

Reviewer 3: We use the `NewValidators` and `DeletedValidators` to determine the final validator set. How is that final validator set ordered, and is the hash of the validator set that is proved by the SNARK dependent on this ordering?

Reviewer 2: Yes, the ordering is important. Maybe there's already a diff message format, but my idea was to begin with a v byte map, where v is the maximum validator set size, each byte of the map tells you what has happened to the validator key that was in that position last time. If the first byte is 0, then that validator key has remained in the same position. If it's 8, then the 9th key from the last validator set is now the first position. If it's 255, then a new public key not present in the last validator set has taken that position. The rest of the diff message is the list of new public keys, in order of appearance. This gives us v bytes + 95n bytes, where n≤v is the number of new validator public keys.

Reviewer 3: When we do validator set diffs for a single epoch transition, i.e. N -> N+1, IIRC the validator set diff implies the order. Given your existing validator set, first you delete all the removed validators, leaving "holes" that are then filled, in order, by the added validators.

However it's not clear to me whether or not that works here

Author: Hm good catch, I'll take a look at how the `ValSetDiff` is calculated:
if the way the valset is updated like how you describe, then so long as the `AddedValidators` produced from epoch X to epoch X + 100 maintains the order of each intermediary epoch (I think it should be? If not, wouldn't be that hard to do) then this should be easy to fit into the existing diff format. Of course will think in more detail to verify this

Reviewer 2: It's one thing to have the protocol preserve the order of validators from epoch to epoch (i.e., if validator A is in position 8 in epoch t, then it remains in 8 in epoch t+1 even if the validators in 7 and/or 9 are replaced), but what if A is removed and re-added to the set? We can't also ensure its position is waiting for it. Hence the permutation/ removal bitmap suggestion.

  DeletedValidators *big.Int // ValSetDiff(FirstEpoch, LastEpoch): bitmap of deleted validators from the `firstEpoch` valset.
}
```

The light clients will need to store verification keys corresponding to the existing versions of the snark circuit, as elaborated in #upgradeability. 

There are two proposed extensions for the `les` protocol:

> Option 1:

Reviewer 3: After coming back to this I'm having trouble understanding why this handles incomplete proof propagation better than approach 2? The only case I've thought of is where this approach is better is when the full node would have proofs that start more than 240 epochs ago, in which case we'd presumably just ask a different full node.

Author: I spoke with @nategraf about this yesterday, one specific case he brought up that I hadn't thought of is the "release process" for new circuits.
Say we upgrade the circuit to v2, this implies a new VK that full nodes must receive in order to validate and accept/propagate further proofs. Depending on how this is implemented, we could be left in a situation where some full nodes have more proofs by nature of having v2 proofs.
The end goal is of course is to allow light clients to verify they are syncing via the fastest available method, so this "proofInventory" is one way, another would be to receive all available proofs from all peers and then follow the shortest path, but that seems even more cumbersome to implement.

```
GetAvailableProofs - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
GetEpochHeadersOrProofs:
  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
    `highestEpoch` is the current highestEpoch the client knows of.
    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
    `requestedProofs`: Set of proofs' metadata requested by the client
  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
  Both responses are subject to message size limits limiting the number of proofs/headers.
EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
```

> Option 2: This option benefits from assumptions on the pattern of proof generation (likely at least one for every ~120 epochs) and by assuming proofs are well distributed (all full nodes have all proofs). The idea here is to pass a bitmap to the `GetEpochHeadersOrProofs` msg that indicates if the client knows the validator set for the past `SIZE_OF_BITMAP` (~240?) epochs. This allows the full node to optimize the set of proofs sent back, while minimizing computation for the client. The main win here would be in a case like this: the client uses a proof to sync to epoch 120, then syncs further to epoch 121, so it knows the validator set of 120 and 121. Then the client goes down for > 120 epochs, and when it comes back the server can figure out to send a proof from 120-240 that we assume is generated, rather than sending 119 epoch headers to verify.

Reviewer 3: I'm not sure that this approach actually relies on those assumptions. The light node gives the full node all the information it needs in order to construct a reasonably efficient batch of proofs/headers to sync the light node.

The light client could include in that request, the maximum number of proofs/headers it's willing to verify.

If the full node cannot meet that request, it returns an empty response. If it *can* meet the request, great!

We can have these values be configurable, such that they can be tightened up towards the optimal values of `(CEIL(epochs / 120), 0)`

Reviewer 2: I'm personally in favor of there being one and only set of proofs and corresponding instance info per epoch that a server will serve. Basically, if clients go offline for a while and get sufficiently far behind, they should just start at the genesis epoch again and verify to the present that way. Maybe after 2 weeks or so downloading the non-SNARK epoch messages becomes more bandwidth/CPU intensive versus just re-bootstrapping with SNARKs. This also very much simplifies proof management these proof serving servers need to do, only storing the minimal number of proofs to get from the genesis to current epoch.

Reviewer 3: Hmm I'm not sure I see why that's necessary. The light client can just ask to sync from the latest epoch N for which it has the validator set where e.g. `N % 120 == 0`.

Reviewer 2: Yes, clarifications below.

Author: Agreed this approach doesn't rely on these assumptions, I was just noting that it benefits from them and produces a more optimal flow if they hold.

Maybe this idea of a max # proofs/headers can solve for the potential problem above of full nodes not having every proof? I'll draw out a potential network flow and think through some data structures for a couple solutions and see if I can decide on a solution

EpochHeadersOrProofs:
  Server sends this back from `GetEpochHeadersOrProofs` with the optimal set of proofs and headers this client can utilize.
```  

It may be preferred to combine both options, and send a bitmap with `handleLogic == false` in option 1.

Regardless of the option chosen, the `les` client must then be modified to request proofs & headers during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`. This will also involve a new function to update the underlying validator set and blockchain data, which will be built to work with checkpoints as well.

Upon receiving proofs and headers in a response message, the light client will verify headers the same way as before, and will verify proofs via a call into `celo-bls-go`'s `verify` fn.

The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.

> ### Upgradeability

Reviewer 3: One thing that might be worth covering is how we will handle gaps in our epoch -> validator set mapping. e.g. what happens if a lightest node wants to check if a transaction was included in a block from a previous epoch for which it doesn't have the validator set? IIRC you solved a similar problem at one rung lower in the validator set > block header > merkle proof hierarchy of the light client protocol.

 think it's fine to not implement the solution for now but it would be nice to know that the message design here will cover that case. (it seems like it does)

Reviewer 2: Sorry, general public, to link a Slack thread you can't see, but some discussion on this in https://celo-org.slack.com/archives/CLPE043GT/p1597444936109200

Reviewer 3: I think that's just about transaction inclusion, right? There are other reasons it's useful to be able to fetch and validate headers, so even if there were other ways to prove transaction inclusion the scheme I described above would still be useful

Reviewer 2: Can you give me some examples of other reasons?

Author: Agreed the plan is for the protocol to be accommodating for this kind of request, but I won't prioritize the strict implementation of these requests yet as you mention

```
GetProofInventory - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
GetEpochHeadersOrProofs:
  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
    `highestEpoch` is the current highestEpoch the client knows of.
    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
    `requestedProofs`: Set of proofs' metadata requested by the client
  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
  Both responses are subject to message size limits limiting the number of proofs/headers.

> EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.

Reviewer 4: Based on our discussion yesterday, I would suggest breaking this message into `GetHeaders/Headers` (which exists) and `GetProofs/Proofs` and only using this message set when the client wants explicit control over what is returned. Option two would use a separate message to share it's sync status and respond with a mixture of `Headers` and `Proofs` messages controlled by the server.

`numEpochs`: Number of epochs we need to sync, the difference between the current epoch of the canonical chain (`currEpoch`) and the client's highest epoch `clientEpoch` = `currEpoch - clientEpoch`

`optimalPath`: For each peer, compute the optimal path of proofs available based on the client's available VKs and known val sets.

`proofRange` : For each proof, the difference between `LastEpoch` and `FirstEpoch`'s indices.

`peerScore`: This is the total number of proofs and headers needed to sync by this peer. For each peer, subtract the range of each proof in `optimalPath` from `numEpochs`, the result is the number of necessary remaining headers. Add `optimalPath`.length to this for the peer's score. 

Whichever peer has the lowest score indicates the shortest path (best set of proofs) for our client. This could also be short-circuited if the optimal score is provided. Once the best peer is chosen, the client requests the set of proofs and headers from this peer that it has assessed as optimal.

Option 2:
```

> GetEpochHeadersOrProofs:

Reviewer 4: We may also need to include information about what proof versions the client is able to accept with this request.

Author: Yes good point

  Client sends (highestEpoch uint, knownValSets *big.Int, maxResponses uint)

>     `knownValSets` is a bitmap of the last ~240 or so epochs that the client knows the val set for

Reviewer 4: I mentioned before, but we should do some thinking about when a bitmap is more or less efficient than a range based representation (i.e. sending a list of `[start, end]` `uint32` pairs that represent that ranges of epochs the light client knows the validator sets for.)

Author: Agreed I'll think more about this and update now that it seems this is the decided (in my mind) path forward

>     `maxResponses` indicates the maximum number of (proofs + headers) the client is willing to accept

Reviewer 4: We should generalize this concept to a "verification cost" metric based on the estimates that @psivesely is discussing, plus a cost metric in terms of data usage, which we could include in a unified number of separate it out.

I am not sure whether this max cost function will work as intended in the case where proofs are well distributed. My thinking is that the client will always want to sync (i.e. it will never decide to stay offline because the cost to sync is too high) so if a server cannot fullfil it's request becasuse the provided max cost is too restrictive, we can either assume that another peer will have a lower cost, and "shop around", or if that is a bad assumtion, the client will simply raise the max cost, having wasted a round trip. As a result, I'm not sure we should include this parameter.

Author: Especially if we're going with option 1 (proof inventory) this added parameter is unnecessary.

If we do add it, though, this idea of a verification cost makes sense to me and is an improvement to this idea.

  Server computes the optimal path of proofs & headers that the client can use to sync fastest and send them back using `EpochHeadersOrProofs`.
EpochHeadersOrProofs:

>   Server sends this back from `GetEpochHeadersOrProofs` with the optimal set of proofs and headers this client can utilize.

Reviewer 4: This can be broken up into `Headers` (which exists) and `Proofs` and sent as a continuous stream of messages. I would suggest this to simplify the network protocol.

Author: Agreed and I'll add this specifically to the design, although I would still like to verify how simply this can be achieved. I agree it simplifies the abstract protocol, but I'm not 100% sure that managing the concurrency of both responses potentially out of sync will be that easy.

# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol

- Date: 2020-07-27

> - Author: @lucasege

Reviewer 5: hi @lucasege thank you for submitting this CIP, I would like you to update it to follow the newly activated CIP process which has it's newly updated template here: https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-template.md

Also, if the CIP is in Draft status, I'd be able to merge the proposal into the repo.

# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol

- Date: 2020-07-27
- Author: @lucasege
- Status: WIP

## Overview

In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.

The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)

## Goals

- Support adding/storing proofs on full nodes

> - Gossip these proofs to propogate them through the network

Reviewer 2: propagate*

- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.

## Proposed Solution

We propose to add two new RPCs: 

> `AddProof(proof []byte, proofId uint)`

Reviewer 6: Are the proofs signed, or can anyone submit a proof and it gets gossiped through the network if the proof is verified?

Author: I think they wouldn't need to be signed, any proof that is verified should be able to be gossiped, allowing us to decentralize the proof creation. C labs will host an initial "proving service" to do this, but community members should be able to contribute.

This is where I would see issues with proof IDs coming in to play

Reviewer 7: Ya.  Agreed about potential issues with proof IDs.  Seems like it's a DOS attack vector.  E.g. a malicious party could submit many duplicated proof but with different proofIDs to fill up all the full nodes' disks.

Author: Yup good points - I plan to key by epoch range like you mentioned @kevjue - specifically `types.PlumoProofEpochs{ firstEpoch, lastEpoch }`

I had originally wanted to follow this path, but thought we might want incremental IDs. But these are good points, and the idea of using a bloom filter makes the incremental IDs unnecessary in my mind

Reviewer 2: Validators should have a fixed list of epoch number pairs (A_1, B_1), (A_2,B_2), ... they accept proofs for. The distance B_i - A_i will be in some fixed set we must determine before carrying out the setup "ceremony." I don't think it makes sense to store (0,32) and (1,33). I think just storing the minimal number of proofs to get from the genesis epoch to the current (or as close as the different fixed lengths let you) is ideal. There's a new circuit @kobigurk has been working on that allows for dummy epoch data and maybe that actually allows for proving an arbitrary number of epochs up to some limit. Hopefully he can chime in here. That would be really nice, as nodes would only have to ceil(t/~120) proofs, where t is the current epoch number.

As for duplicate proofs, nodes should probably just store the first valid proof they receive for a given instance (epoch number pair).

> `AddProof(proof []byte, proofId uint)`

Reviewer 7: Will this RPC call verify that the plumo proof is valid?  Or does that even need to be done? (I don't know enough about the plumo/SNARK proofs to know if that's even a valid question).

Author: Yes in my mind this could and should be done. This would help prevent potential for DDoS via someone spamming bad proofs to one node which then propagate to the whole network. I'll add this to the CIP after I finish synthesizing the rest of the feedback

and

> `Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.

Reviewer 6: How many bytes do we expect the proofs to be?

Author: Depends on many factors, but the estimates I've seen for what we may use range from ~5kB - ~12kB. 
The dependent factors include (probably not exhaustive): max number of epochs the circuit is designed for and the type of proof protocol (marlin vs. groth16)

Reviewer 2: We've decided on Groth16 proofs. Marlin and Groth16 both have fully succinct proofs, so the proof size is independent of the number of epochs the circuit is designed for. Groth16 proofs over BW6-761 will be 764B. You will also need to send the client half the instance. The instance includes the public keys at epochs A and B, where B is some number of epochs later. The client will already have validator keys for epoch A and needs the keys at epoch B. 100 public keys over BLS12-377 are 18.9KB.

This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.

To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
```

> PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`

Reviewer 7: Should there be a new message type that will broadcast the new proof's metadata.  E.g. the proof's ID/hash and epoch range?  Analogous to `NewBlockHashes` and `NewPooledTransactionHashes` (https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01).

Author: Very good point, I'll add this to the CIP, wasn't aware of how exactly this mechanism worked

> PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`

Reviewer 7: Seems like there could be plumo proofs that have overlapping epoch ranges.  In that case, are all the full nodes expected to store all of the proofs, or will it prune redundant proofs to minimize disk space?

Author: This is certainly a concern - Currently I was planning to just store all proofs for simplicity, although pruning redundant proofs wouldn't add much burden. Proofs aren't very large (estimated between 5-12kB) so I don't think this is necessary, but please let me know what you think.

> GetPlumoProofMsg - Request plumo proof(s) from a peer

Reviewer 6: Under what condition would a `GetPlumoProofMsg` be sent to all its peers?
Would be good to have a little more detail here on how exactly the gossiping will work

Author: The goal of this msg `GetPlumoProofMsg` is to allow new syncing nodes to fetch previous proofs, and to allow nodes that have gone down to fetch proofs they may have missed.

I'm not sure exactly where this should be handled, but I'm guessing within `downloader`, when a node is syncing, it would send this `get` message.

In regard to sending to all its peers, I imagined sending one message to each peer you know of, and then just taking the union set of all proofs from those responses as your set (in case any peer is similarly out of sync). Maybe this is overkill, if for example peers only respond with `PlumoProofMsg`s if they are in sync already.

Does this address what you were getting at? I'm not very familiar with the already implemented gossiping protocols so appreciate feedback and advice

```
`GetPlumoProofMsg` will be implemented by having the requester send a set of proof IDs that they hold, and ask for any they are missing. This may be optimized using a bloom filter as well.

> The `les` protocol must then also be extended via at least two messages:

Reviewer 4: Implementation note: `les/4` is pending on `master` to include gateway fee changes that are not fully complete. We will want to provide MVP support for those (which may even be a stub) or figure out another way to handle this if we want to have these modification as part of `les/4` (Another option is to move gateway fee messages to `les/5` as well if there is no time to get it fully ready)

Author: Hm good point, thanks for pointing this out. I'm inclined to go the second route (plumo = les/4, gateway fees to les/5), let me know if you agree or if this changes anything else @nategraf

```

> GetEpoch - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.

Reviewer 6: Slightly confused by the wording here and line 39, so the `GetEpoch` is a message sent by a client and then the server responds with `Epochs`?

Author: Yes, I imagine this working very similarly to existing `les` messages, such as `GetBlockHeadersMsg` and `BlockHeadersMsg` where I believe the client sends `GetBlockHeadersMsg` with some corresponding parameters (block number/hash) to a full node, and the full node sends back a `BlockHeadersMsg` with the same parameters for the light client to consume.
I'd imagine `GetEpochsMsg` should work similarly, where the client sends this msg with the parameter of the client's highest known epoch, and the full node responds with `EpochsMsg` with the same parameters and the corresponding set of headers/proofs to progress the client's sync

> GetEpochsMsg - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.

Reviewer 7: nit:  Consider renaming to `GetCertifiedEpochsValidatorSet`, or something along those lines.

> EpochsMsg - Receives a mix of header(s) and proof(s) from the server.

Reviewer 4: I would propose that that instead of this extension we add messages that will allow the client to be inform of what proofs the server can provide filtered by block range, and to download those proofs. That would give the client flexibility to decide what mix of proofs and headers it wants to use for syncing.

Author: hm - I think this is something worth discussing synchronously. I'm personally partial to the original plan, because it seems to reduce any necessary back/forth between client/server.
I don't see any uses for the client being able to chose the mix of proofs/headers, is there a benefit you see?
The motivation behind the original approach was that the light client doesn't care _how_ it syncs, only that it syncs in the fastest way. It seems logical imo for the server (with all relevant info, no need to download more) to interpret this and return the data necessary to make this happen

```

The `les` client must then be modified to send a `GetEpochsMsg` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.

Upon receiving the `EpochsMsg` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.

> The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.

Reviewer 4: Could you elaborate on this a bit? I don't quite understand from this description.

Author: This is an area I don't have great context on, I was tipped off that this will likely be an issue by asa.

My understanding is that the underlying `lightchain` data is likely not readily accessible for the light client when syncing in `downloader` so there will need to be changes to expose this information and may require new methods to update the validator set

Reviewer 3: I think what you might be talking about is that we need some way to be able to write an Istanbul validator snapshot for a given epoch after a proof is verified. Currently, I believe the only way for light clients to write a snapshot is to validate an epoch header.

Author: Yup exactly, that's my understanding of the necessary changes

## Alternative Solutions

`GetPlumoProofMsg` could also request *all* plumo proofs from the receiver, leaving the responsibility of filtering seen proofs with the caller. This is a simpler implementation, but is less scalable as we increase the amount of stored proofs. 

On the type of ID to use for proofs: an incrementing counter seems most "user-friendly" because of the ability to sense gaps, but could introduce some synchronization issues in the future.

Specific aspects of the `les` changes may have tradeoffs that I've yet to encounter, so any input and feedback on things to consider there would be helpful.

## Risks

> Altering the `eth` protocol introduces some risk of introducing bugs.

Reviewer 4: Just some added clarification on the risk.
```suggestion
Altering the `eth` protocol introduces some risk of introducing bugs, which could inhibit a nodes ability to connect to the network.
```

> Introducing bugs to mobile client syncing.

Reviewer 4: ```suggestion
Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain.
```

## Proposed Solution

We propose to add two new RPCs: 
* `AddProof(proof []byte, firstEpoch uint, lastEpoch uint, versionNumber uint)`
* `Proofs() ([][]byte)`.

`versionNumber` here refers to the snark circuit version, for more info refer to #upgradeability below.

When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propagate the proof throughout the network.

This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by it's combination of first and last epoch. 

To propagate these proofs throughout the network, we also propose to extend the `eth` protocol with three new message types: 
```
NewPlumoProof (0x019)
[[firstEpoch, lastEpoch, versionNumber], ...]
Specify one or more new plumo proofs which have appeared on the network.
This is intended to be honored with a proceeding GetPlumoProofsMsg to request the proof.

GetPlumoProofsMsg (0x1a)
[[firstEpoch, lastEpoch, versionNumber], ...]

> Sent from node A to node B with the set of proofs that node A currently has, requires node B to return a PlumoProofsMsg.

Reviewer 4: In steady state, sending the `GetPlumoProodsMsg` with the list of proofs the node has will be much larger than sending a light of the proofs the node wants. Especially for use by long-running nodes who send `GetPlumoProofs` in response to a `NewPlumoProof` message indicating the existence of a proof the node does not have, it will be more efficient to send a message containing only the identifier for the proof the node wants. Adding a flag, or other indicator in the request to allow for requesting specific proof(s) would be beneficial.

Another case where receiving every proof the requestor does not have may not be ideal is if the requester is only interested in a certain subset of proofs (e.g. proofs with the latest version number, or proofs after a certain epoch)

Author: Yes certainly, I was wondering if I should split this message into two messages, one to request specifics and one to request any you may be missing.

I like the idea of including this in one message since the logic is of course very similar. 

I'll adapt the spec to follow this idea ^ thanks

This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
```

The `les` protocol must then also be extended via at least two messages:
```
GetCertifiedEpochsValidatorSet - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
CertifiedEpochsValidatorSet - Receives a mix of header(s) and proof(s) from the server.
```

The `les` client must then be modified to send a `GetCertifiedEpochsValidatorSet` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.

Upon receiving the `CertifiedEpochsValidatorSet` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.

The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.

### Upgradeability

> In order to support an upgradeable snark circuit, we utilize a `versionNumber` parameter with every proof. This corresponds to a version of the snark circuit and thus a corresponding verification key. These version numbers will be stored on chain, within the `blockchain_parameters` smart contract, with the corresponding hash of the verification key. The clients are expected to store all applicable verification keys.

Reviewer 4: Doesn't storing the version numbers on-chain create a chicken-and-egg problem in that the user of a plumo proof is not synced to the chain yet? It may be necessary to ship verification keys in the client code instead.

Author: This is a great point - I was thinking we could ship the VKs in the client code, but then have the clients verify each key against the stored reference hash on chain, but of course if you aren't synced you won't see that hash.

I agree, I thought through some plans to allow for this, but if we want to use chain state to verify keys then I believe you need to set boundaries for verification keys, i.e. you can only use version n to prove transitions on epochs greater than X (since you publish the hash to the chain in epoch X, can't access it before). I think this is worse than just shipping all VKs with the client like you mentioned, since then we can upgrade our circuit and use it for any epochs, allowing us to replace a potentially inferior proof scheme.

For any uses of the snark's `verify` fn in the above protocols, a client is expected to:
* Grab the verification key corresponding to the proof's `versionNumber`
* Hash this key and compare the result to `blockchain_parameters` published VK hash.
* If equal, the client may proceed to verify the proof.

This allows the snark circuit to be upgraded via incrementing future proofs' `versionNumber` and publishing a new mapping of `versionNumber: VK_hash`, as well as distributing the new VK.

### Bad proofs
Bad proofs should be prevented from propagated throughout the network via the `verify` check that each node performs. In the case that any participant receives a proof that does not pass verification for its inputs, that participant is expected to drop the sender. 

## Alternative Solutions

> Alternatively, `GetPlumoProofMsg` could utilize a bloom filter instead, allowing for a constant sized message. However, bloom filters allow for false positives: a false positive here would indicate to a peer that the sender has a proof when the sender does not. This could prevent certain proofs from ever getting to some nodes. For example, say node A is on the network and has received a certain amount of proofs via traditional propagation. Node A then goes down for a short period, and within that period proof P is added and propagated. Node A then comes back online, and sends a `GetPlumoProofsMsg` to request any proofs it may have missed, and sends along the corresponding bloom filter for the proofs it holds. If proof P happens to be a false positive for the set of proofs node A already has, then all of node A's peers will simply return nothing. Thus node A will never receive proof P. This can then propagate similarly through the network either through a similar path as node A, or for any nodes where all of their peers have already "lost" proof P.

Reviewer 4: 👍 One place where bloom filters could be handy is if the proof serving peer provides a filter to help answer the client peer's question of "does this peer have a proof with identifier X". That is not how the protocol is specified here though, so I agree that it doesn't seem relevant here.
